crates/shielded_token/src/storage_key.rs

```diff
@@ -85,6 +89,29 @@ pub fn masp_last_inflation_key<TransToken: trans_token::Keys>(
         .with_segment(MASP_LAST_INFLATION_KEY.to_owned())
 }
 
+/// Check if the given storage key is a masp reward balance key
+pub fn is_masp_undated_balance_key(key: &storage::Key) -> Option<Address> {
+    match &key.segments[..] {
+        [
+            DbKeySeg::AddressSeg(address::MASP),
+            DbKeySeg::StringSeg(prefix),
+            DbKeySeg::StringSeg(token),
+        ] if prefix == MASP_UNDATED_BALANCE_KEY => {
+            Address::from_str(token).ok()
+        }
+        _ => None,
+    }
+}
+
+/// Obtain the storage key for the undated balance of a token
+pub fn masp_undated_balance_key(token_address: &Address) -> storage::Key {
+    storage::Key::from(address::MASP.to_db_key())
+        .push(&MASP_UNDATED_BALANCE_KEY.to_owned())
+        .expect("Cannot obtain a storage key")
+        .push(&token_address.to_string().to_db_key())
+        .expect("Cannot obtain a storage key")
```

Comment:
replace the `.push(...).expect(...)` pattern with `.with_segment(...)`

Reply:
Oops, did not realize this function could be used here. Fixed, thanks!